Optimize if range check to replace jumps to bit operation

JulieLeeMSFT opened the pull request:
It detects if statement with specifc range check patterns and optimizes them to use bit operation instead of jumps for each pattern.

Pattern 1 optimized in this PR:
```c#
    static bool IsReservedPattern(int a)
    {
        if (a == 97 || a == 98 || a == 100 || a == 102)
            return true;
        return false;
    }
```

Before:
```asm
    cmp      ecx, 97
    je       SHORT G_M14484_IG04
    cmp      ecx, 98
    je       SHORT G_M14484_IG04
    cmp      ecx, 100
    je       SHORT G_M14484_IG04
    cmp      ecx, 102
    jne      SHORT G_M14484_IG06
G_M14484_IG04:    
    mov      eax, 1
    ret      
G_M14484_IG06:    
    xor      eax, eax
    ret      
```

After:
```asm
    sub      ecx, 97
    cmp      ecx, 5
    ja       SHORT G_M14484_IG05
    mov      eax, 43
    bt       eax, ecx
    jae      SHORT G_M14484_IG05
    mov      eax, 1
    ret      
G_M14484_IG05:  
    xor      eax, eax
    ret  
```
Contributes to #8418.

ewhapdx committed:
Optimize if range check to bit operation by changing if tree to switch tree.

JulieLeeMSFT commented:
Reference: https://egorbo.com/llvm-range-checks.html

ewhapdx committed:
[if range check opt] handle unsorted patterns.

ewhapdx committed:
[if range check opt] Handle appending a statement for side effects.

danmoseley commented:
@JulieLeeMSFT  you can set the code blocks to colorize, I modified your top post, basically just put C# or asm immediately after the first three `

danmoseley commented:
It's an interesting trick we can use in c# and the range doesn't have to be contiguous (!), simply no more than register width between highest and lowest values. @Stephentoub does Regex take advantage of this already? Perhaps it's an element of how searching is already optimized?

danmoseley commented:
@MihaZupan  but assuming this is already well known.

stephentoub commented:
> @stephentoub does Regex take advantage of this already? Perhaps it's an element of how searching is already optimized?

In concept.  It generates a branch-free version, e.g. for the set `[abdf]` that's being checked for in IsReservedPattern above, it emits the equivalent of this:

```C#
static bool IsReservedPattern(int a)
{
    uint charMinusLowUInt32;
    return ((int)((0xD4000000U << (short)(charMinusLowUInt32 = (ushort)(a - 'a'))) & (charMinusLowUInt32 - 32)) < 0);
}
```

```asm
C.IsReservedPattern(Int32)
    L0000: add ecx, 0xffffff9f
    L0003: movzx eax, cx
    L0006: movsx rdx, ax
    L000a: mov ecx, 0xd4000000
    L000f: shlx edx, ecx, edx
    L0014: add eax, 0xffffffe0
    L0017: and eax, edx
    L0019: shr eax, 0x1f
    L001c: ret
```

The code for this in the source generator is here:
https://github.com/dotnet/runtime/blob/05e25ff8c3a623df74720f430c270eba0f1d41b0/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L4857-L4920

JulieLeeMSFT commented:
> @JulieLeeMSFT you can set the code blocks to colorize, I modified your top post, basically just put C# or asm immediately after the first three `

Yes, I tested the code for non contiguous values and unsorted values. It works now. I am checking some other details.

ewhapdx committed:
[if range check opt] Code clean-up.

EgorBo commented:
> It generates a branch-free version, e.g. for the set [abdf] that's being checked for in IsReservedPattern above, it emits the equivalent of this:

gcc emits closer to what Julie does: https://godbolt.org/z/rj1cPYz1a, basically, a single jump to filter out obviously wrong values and then a bit test (bt). No idea what's faster in reality. Although, it's not related to this PR directly, because what it does is it just emits a SWITCH opcode instead of series of comparisons

EgorBo commented on src/coreclr/jit/optimizebools.cpp:
```diff
@@ -1570,3 +1570,698 @@ PhaseStatus Compiler::optOptimizeBools()
+                        {
+                            auto leftCondition1 = condition1->gtGetOp1(); // op1 of condition1 from currBb
+                            auto leftCondition2 = condition2->gtGetOp1(); // op1 of condition2 from prevBb
+                            if (leftCondition1->IsLocal() &&
```
I don't think this is a correct check - better make it more conservative and check `OperIs(GT_LCL_VAR)`
because in your case a struct field can pass this check but two locals will have different offsets

EgorBo commented on src/coreclr/jit/optimizebools.cpp:
```diff
@@ -1570,3 +1570,698 @@ PhaseStatus Compiler::optOptimizeBools()
+        unsigned int btPatterns = 0;
+        for (int idxPattern = 0; idxPattern < patternCount; idxPattern++)
+        {
+            btPatterns |= (1 << (optRngPattern.optGetPattern(idxPattern) - minPattern));
```
I am not sure I understand why do you need to calculate `btPatterns` here - I though the idea is just to emit a SWITCH and let the jit handle it, you only need to make sure switch has less than 64 cases (or some other profitability checks)

JulieLeeMSFT replied:
Per offline discussion, skipped matching blocks with side effects to allow the change of block order. And, removed bitmap.

JulieLeeMSFT commented on src/coreclr/jit/optimizebools.cpp:
```diff
@@ -1570,3 +1570,698 @@ PhaseStatus Compiler::optOptimizeBools()
+    }
+
+    // Change from GT_JTRUE to GT_SWITCH
+    rootTree->ChangeOper(GT_SWITCH, GenTree::PRESERVE_VN);
```
Removed.

ewhapdx committed:
[if range check opt] Use gtUpdateStmtSideEffects to update flags after changing statement and trees.

ewhapdx committed:
[if range check opt] Handle MIN and MAX limits for pattern values.

JulieLeeMSFT commented:
**Other patterns tested:**

**Pattern 2:** handles a tree with GT_COMMA. 

```C#
        if (c == 'a' || c == 'b' || c == 'd' || c == 'f')
        {
            return true;
        }
        return false;
```

**Pattern 3**: handles unsorted patterns.
```C#
        if (a == 102 || a == 97 || a == 98 || a == 100)
            return true;
        return false;
``` 

**Pattern 4**: handles min and max limit.
```C#
        if (a == -9223372036854775806 || a == -9223372036854775807 || a == -9223372036854775804)
            return true;
        return false;
        
        if (a == 9223372036854775806 || a == 9223372036854775807 || a == 9223372036854775804)
            return true;
        return false;
```

ewhapdx committed:
[if range check opt] Skip optimization for a tree with side effects.

ewhapdx committed:
[if range check opt] Don't optimize if block is marked with BBF_DONT_REMOVE.

ewhapdx committed:
[if range check opt] Optimize GT_CNS_INT patterns only.

ewhapdx committed:
[if range check opt] Skip cases that Lowering does not convert to a bit test.

ewhapdx committed:
[if range check opt] Skip if switch cases or default don't jump to switch->bbNext.

ewhapdx committed:
Change Switch conversion to happen after Optimize Layout to avoid Switch and bbNext blocks get separated.

JulieLeeMSFT commented:
Hi @EgorBo, this is ready for review. CC @dotnet/jit-contrib.
  
Asmdiff shows size regression, but they are all converted to a bit test instead of multiple compares and jumps.
- aspnet
![image](https://github.com/dotnet/runtime/assets/63486087/6a503a57-bc4c-4e54-bd43-56b261837ef1)

- benchmarks: 
![image](https://github.com/dotnet/runtime/assets/63486087/f162b65e-89f8-4b6d-9aee-deaf494c6278)

JulieLeeMSFT commented:
CI test shows 20K size regressions on x64 platforms vs 5K on arm64. I will check on those regressions.

EgorBo commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5048,6 +5048,10 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+
+        // Conditional to Switch conversion
+        //
+        DoPhase(this, PHASE_Find_Specific_Pattern, &Compiler::optFindSpecificPattern);
```
I assume it was a temporal name for the phase, right? needs to be renamed to something like optSwitchRecognition + PHASE_SWITCH_RECOGNITION or something like that

EgorBo commented on src/coreclr/jit/optimizebools.cpp:
```diff
@@ -1570,3 +1570,1008 @@ PhaseStatus Compiler::optOptimizeBools()
+//
+BasicBlock* OptRangePatternDsc::optGetDefaultJmpBB()
+{
+    return m_defaultJmpBB;
```
it feels like your class has many very simple methods/accessors which bloat the code size - maybe consider replacing simple ones like this with just accessing `m_defaultJmpBB` directly?

JulieLeeMSFT commented:
Checked the size regression.
When converting 2 or 3 conditionals, the bit test code is about 10 bytes larger and perf score is about 1-2 score higher. 

When it becomes 4 conditional conversion, it shows size reduction and perf score improvements:
![image](https://github.com/dotnet/runtime/assets/63486087/251f04bc-f9d4-4a02-b177-b89305c9f965)

ewhapdx committed:
Refactored the code for Switch Recognition. Match pattern for >=3 conditionals.

JulieLeeMSFT commented:
I found out that not all Switch basic blocks are converted to a bit test in Lowering phase. So, added checks to skip Switch conversion for the cases that Lowering will not perform bit test conversion. 

- TryLowerSwitchToBitTest performs the BT conversion for certain cases only. 
  - Switch jump targets need to be 2, and it cannot be more than 3. 
  - One of jump targets needs to be `Switch->bbNext`. If they are not adjacent, it is not converted to a BT. So, in Optimize Layout phase when Switch or its bbNext block gets reordered and they are not adjacent, it fails to convert ot a BT. So, moved the `Switch Recognition` phase to after the `Optimize Layout` phase.
```
    // One of the case blocks has to follow the switch block. This requirement could be avoided
    // by adding a BBJ_ALWAYS block after the switch block but doing that sometimes negatively
    // impacts register allocation.
```

ewhapdx committed:
Refactored Switch recognition.

JulieLeeMSFT commented:
> If this optimization is only for bit tests, then how much harder would it be to change things to a bit test (`cns & (1 << val)`) directly at this point? It seems like it would be much clearer instead of trying to depend on lowering to further optimize a switch.

I have a code to get a bitmap value from patterns in my initial commit, so it is not difficutl to convert it to AND and Shift tree. When I worked on this PR, I didn't expect that there would be such limitations in lowering. 
@EgorBo, what do you think?

JulieLeeMSFT commented:
> If this optimization is only for bit tests, then how much harder would it be to change things to a bit test (`cns & (1 << val)`) directly at this point? It seems like it would be much clearer instead of trying to depend on lowering to further optimize a switch.

Discussed with EgorBo offline. We want to keep Switch conversion so that we can extend it to handle normal jump tables without bt in the future.

ewhapdx committed:
Switch conversion build error fix.

ewhapdx committed:
Switch conversion optimization: build error fix for an unused variable.

ewhapdx committed:
Switch conversion optimization: fortmat fix.

JulieLeeMSFT commented:
By limiting the pattern matching to 3 or more BBJ_COND blocks, the size regression has reduced significantly to an acceptable level.
CI error are known issues.

![image](https://github.com/dotnet/runtime/assets/63486087/11e6ce84-356d-4ef1-a18f-95b2abb75815)

![image](https://github.com/dotnet/runtime/assets/63486087/fb3a854b-a1d0-4e7d-b057-c4923c3164a7)

AndyAyersMS commented:
> By limiting the pattern matching to 3 or more BBJ_COND blocks, the size regression has reduced significantly to an acceptable level.

This may be a little misleading, 50% of the method contexts fail to replay (likely as a result of #88163). I kicked off a new SPMI collection so you might want to merge this up tomorrow and try remeasuring.

JulieLeeMSFT committed:
Switch recognition: change phase text name.

JulieLeeMSFT commented:
Reran the diffs with a new superpmi collection. The code size is 950B regression, which I think is an acceptable level.
@EgorBo, it is ready for review. CC @dotnet/jit-contrib.
 
<img width="664" alt="image" src="https://github.com/dotnet/runtime/assets/63486087/5fad5d66-c28e-414c-b900-3201f34ff242">

JulieLeeMSFT committed:
Add test cases for Switch recognition.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into 8418

JulieLeeMSFT commented:
Waiting for EgorBo's review.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into 8418

EgorBo committed:
Move to a separate file

EgorBo committed:
Code clean up

EgorBo committed:
More clean up

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into 8418

EgorBo committed:
Refactoring

EgorBo commented:
/azp list

EgorBo commented:
/azp run runtime-coreclr outerloop, runtime-coreclr pgo

EgorBo committed:
Update switchrecognition.cpp

EgorBo committed:
Update switchrecognition.cpp

EgorBo committed:
Optimization

EgorBo committed:
Clean up

EgorBo committed:
Ignore cold blocks

EgorBo committed:
Disable for arm, no meaningful diffs anyway

EgorBo commented:
/azp run runtime-coreclr outerloop, runtime-coreclr pgo

EgorBo commented:
Slightly refactored the impl to make it smaller and extendable in future, so far it finds ~4000 contexts to improve.

![image](https://github.com/dotnet/runtime/assets/523221/de6050cc-15ca-4dd0-9456-a1ea9a984ed7)

Regressions also look like perf improvements because of less branches. We can increase minimal number of tests from 3 to 4 to get rid of size regressions but I guess we'd better do it based on microbenchmarks reports as the overall diffs aren't bad.

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into 8418

teo-tsirpanis commented:
(changing milestone to 9.0.0, I assume it won't be backported)

JulieLeeMSFT commented:
Verified that `char`, `byte`, and `short` types generate `bt` instruction. 
cc @stephentoub. 

- `char` type
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool TestRange(char c)
    {
        if (c == ' ' || c == '\t' || c == '\r' || c == '\n')
            return true;
        return false;
    }
```

```asm
IN0001: 000000 movzx    rcx, cx
IN0002: 000003 cmp      ecx, 32
IN0003: 000006 je       SHORT G_M24766_IG07

IN0004: 000008 cmp      ecx, 13
IN0005: 00000B ja       SHORT G_M24766_IG05

IN0006: 00000D mov      eax, 0x19FF
IN0007: 000012 bt       eax, ecx
IN0008: 000015 jae      SHORT G_M24766_IG07

G_M24766_IG05:        ; offs=0x000017, size=0x0002, bbWeight=0.50, PerfScore 0.12, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB04 [0005], byref
IN0009: 000017 xor      eax, eax
IN000b: 000019 ret

G_M24766_IG07:        ; offs=0x00001A, size=0x0005, bbWeight=0.50, PerfScore 0.12, gcVars=0000000000000000 {}, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB05 [0004], gcvars, byref
IN000a: 00001A mov      eax, 1
IN000c: 00001F ret
```

- `short` type
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool TestRange(short c)
    {
        if (c == ' ' || c == '\t' || c == '\r' || c == '\n')
            return true;
        return false;
    }
```

```asm
IN0001: 000000 movsx    rcx, cx
IN0002: 000004 cmp      ecx, 32
IN0003: 000007 je       SHORT G_M54667_IG07

IN0004: 000009 cmp      ecx, 13
IN0005: 00000C ja       SHORT G_M54667_IG05

IN0006: 00000E mov      eax, 0x19FF
IN0007: 000013 bt       eax, ecx
IN0008: 000016 jae      SHORT G_M54667_IG07

G_M54667_IG05:        ; offs=0x000018, size=0x0002, bbWeight=0.50, PerfScore 0.12, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB04 [0005], byref
IN0009: 000018 xor      eax, eax
IN000b: 00001A ret

G_M54667_IG07:        ; offs=0x00001B, size=0x0005, bbWeight=0.50, PerfScore 0.12, gcVars=0000000000000000 {}, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB05 [0004], gcvars, byref
IN000a: 00001B mov      eax, 1
IN000c: 000020 ret
```

- `byte` type
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool TestRange(byte c)
    {
        if (c == ' ' || c == '\t' || c == '\r' || c == '\n')
            return true;
        return false;
    }
```

```asm
IN0001: 000000 movzx    rcx, cl
IN0002: 000003 cmp      ecx, 32
IN0003: 000006 je       SHORT G_M29734_IG07

IN0004: 000008 cmp      ecx, 13
IN0005: 00000B ja       SHORT G_M29734_IG05

IN0006: 00000D mov      eax, 0x19FF
IN0007: 000012 bt       eax, ecx
IN0008: 000015 jae      SHORT G_M29734_IG07

G_M29734_IG05:        ; offs=0x000017, size=0x0002, bbWeight=0.50, PerfScore 0.12, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB04 [0005], byref
IN0009: 000017 xor      eax, eax
IN000b: 000019 ret

G_M29734_IG07:        ; offs=0x00001A, size=0x0005, bbWeight=0.50, PerfScore 0.12, gcVars=0000000000000000 {}, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB05 [0004], gcvars, byref
IN000a: 00001A mov      eax, 1
IN000c: 00001F ret
```

JulieLeeMSFT commented:
if range values are contiguous, it removes `switch` and simplifies to fall through to its unique successor with BBJ_ALWAYS. So, `bt` is not generated.

stephentoub commented:
Thanks, @JulieLeeMSFT. FYI, it fails to do so if you don't use `return true;`, e.g.
```C#
static bool TestRange(char c) => c == ' ' || c == '\t' || c == '\r' || c == '\n';
```

JulieLeeMSFT commented:
I will check on it.

> Thanks, @JulieLeeMSFT. FYI, it fails to do so if you don't use `if / return true;`, e.g.
> 
> ```cs
> [MethodImpl(MethodImplOptions.NoInlining)]
> static bool TestRange(char c) => c == ' ' || c == '\t' || c == '\r' || c == '\n';
> ```

JulieLeeMSFT commented:
`char` range value creates a COMMA with STORE_LCL_VAR, which we skip for switch conversion. Then it leaves only 2 JTRUE nodes, one for `\t` and the other for `\r`, which is less than the minimum number of BBs we require for switch conversion. So, optimization does not kick in.
For `char`, if we add one more condition, `!`, the optimization kicks in and generates `bt` as shown below. 
Verified that short and byte also work when adding one more value, `!`.
cc @stephentoub. 

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static bool TestRange(char c) => c == ' ' || c == '\t' || c == '\r' || c == '\n' || c == '!';
```

```asm
***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x003 )
N008 (  8,  9) [000003] -A---------                         *  JTRUE     void   $VN.Void
N007 (  6,  7) [000002] JA-----N---                         \--*  EQ        int    $101
N005 (  4,  5) [000030] -A---------                            +--*  COMMA     int    $100
N003 (  3,  4) [000028] DA---------                            |  +--*  STORE_LCL_VAR int    V02 cse0         d:1 $VN.Void
N002 (  3,  4) [000022] -----------                            |  |  \--*  CAST      int <- ushort <- int $100
N001 (  2,  2) [000000] -----------                            |  |     \--*  LCL_VAR   int    V00 arg0         u:1 $80
N004 (  1,  1) [000029] -----------                            |  \--*  LCL_VAR   int    V02 cse0         u:1 $100
N006 (  1,  1) [000001] -----------                            \--*  CNS_INT   int    32 $43

------------ BB02 [0001] [005..00A) -> BB06(0.5),BB03(0.5) (cond), preds={BB01} succs={BB03,BB06}

***** BB02 [0001]
STMT00002 ( 0x005[E-] ... 0x008 )
N004 (  5,  5) [000009] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000008] J------N---                         \--*  EQ        int    $102
N001 (  1,  1) [000031] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000007] -----------                            \--*  CNS_INT   int    9 $44

------------ BB03 [0002] [00A..00F) -> BB06(0.5),BB04(0.5) (cond), preds={BB02} succs={BB04,BB06}

***** BB03 [0002]
STMT00003 ( 0x00A[E-] ... 0x00D )
N004 (  5,  5) [000013] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000012] J------N---                         \--*  EQ        int    $103
N001 (  1,  1) [000032] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000011] -----------                            \--*  CNS_INT   int    13 $45

------------ BB04 [0003] [00F..014) -> BB06(0.5),BB05(0.5) (cond), preds={BB03} succs={BB05,BB06}

***** BB04 [0003]
STMT00004 ( 0x00F[E-] ... 0x012 )
N004 (  5,  5) [000017] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000016] J------N---                         \--*  EQ        int    $104
N001 (  1,  1) [000033] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000015] -----------                            \--*  CNS_INT   int    10 $42

------------ BB05 [0004] [014..01A) (return), preds={BB04} succs={}

***** BB05 [0004]
STMT00005 ( 0x014[E-] ... 0x019 )
N004 (  7,  4) [000021] -----------                         *  RETURN    int    $VN.Void
N003 (  6,  3) [000020] -----------                         \--*  EQ        int    $105
N001 (  1,  1) [000034] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000019] -----------                            \--*  CNS_INT   int    33 $47

------------ BB06 [0005] [01A..01C) (return), preds={BB01,BB02,BB03,BB04} succs={}

***** BB06 [0005]
STMT00001 ( 0x01A[E-] ... 0x01B )
N002 (  2,  2) [000005] -----------                         *  RETURN    int    $VN.Void
N001 (  1,  1) [000026] -----------                         \--*  CNS_INT   int    1 $46
```


```asm
IN0001: 000000 movzx    rcx, cx
IN0002: 000003 cmp      ecx, 32
IN0003: 000006 je       SHORT G_M24766_IG05

IN0004: 000008 cmp      ecx, 13
IN0005: 00000B ja       SHORT G_M24766_IG07

IN0006: 00000D mov      eax, 0x19FF
IN0007: 000012 bt       eax, ecx
IN0008: 000015 jb       SHORT G_M24766_IG07

G_M24766_IG05:         

IN0009: 000017 mov      eax, 1
IN000d: 00001C ret

G_M24766_IG07:       
IN000a: 00001D cmp      ecx, 33
IN000b: 000020 sete     al
IN000c: 000023 movzx    rax, al
IN000e: 000026 ret
```

JulieLeeMSFT commented:
BBs for 

```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool TestRange(char c)
    {
        if (c == ' ' || c == '\t' || c == '\r' || c == '\n')
            return true;
        return false;
    }
```


```asm
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight   [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1    [000..005)-> BB06(0.5),BB02(0.5)     ( cond )                     i
BB02 [0001]  1       BB01                  0.50 [005..00A)-> BB06(0.5),BB03(0.5)     ( cond )                     i
BB03 [0002]  1       BB02                  0.50 [00A..00F)-> BB06(0.5),BB04(0.5)     ( cond )                     i
BB04 [0003]  1       BB03                  0.50 [00F..014)-> BB05(0.5),BB06(0.5)     ( cond )                     i
BB05 [0005]  1       BB04                  0.50 [016..018)                           (return)                     i
BB06 [0004]  4       BB01,BB02,BB03,BB04   0.50 [014..016)                           (return)                     i
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [0000] [000..005) -> BB06(0.5),BB02(0.5) (cond), preds={} succs={BB02,BB06}

***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x003 )
N008 (  8,  9) [000003] -A---------                         *  JTRUE     void   $VN.Void
N007 (  6,  7) [000002] JA-----N---                         \--*  EQ        int    $101
N005 (  4,  5) [000028] -A---------                            +--*  COMMA     int    $100
N003 (  3,  4) [000026] DA---------                            |  +--*  STORE_LCL_VAR int    V02 cse0         d:1 $VN.Void
N002 (  3,  4) [000020] -----------                            |  |  \--*  CAST      int <- ushort <- int $100
N001 (  2,  2) [000000] -----------                            |  |     \--*  LCL_VAR   int    V00 arg0         u:1 $80
N004 (  1,  1) [000027] -----------                            |  \--*  LCL_VAR   int    V02 cse0         u:1 $100
N006 (  1,  1) [000001] -----------                            \--*  CNS_INT   int    32 $43

------------ BB02 [0001] [005..00A) -> BB06(0.5),BB03(0.5) (cond), preds={BB01} succs={BB03,BB06}

***** BB02 [0001]
STMT00002 ( 0x005[E-] ... 0x008 )
N004 (  5,  5) [000009] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000008] J------N---                         \--*  EQ        int    $102
N001 (  1,  1) [000029] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000007] -----------                            \--*  CNS_INT   int    9 $44

------------ BB03 [0002] [00A..00F) -> BB06(0.5),BB04(0.5) (cond), preds={BB02} succs={BB04,BB06}

***** BB03 [0002]
STMT00003 ( 0x00A[E-] ... 0x00D )
N004 (  5,  5) [000013] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000012] J------N---                         \--*  EQ        int    $103
N001 (  1,  1) [000030] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000011] -----------                            \--*  CNS_INT   int    13 $45

------------ BB04 [0003] [00F..014) -> BB05(0.5),BB06(0.5) (cond), preds={BB03} succs={BB06,BB05}

***** BB04 [0003]
STMT00004 ( 0x00F[E-] ... 0x012 )
N004 (  5,  5) [000017] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  3) [000016] N------N-U-                         \--*  NE        int    $104
N001 (  1,  1) [000031] -----------                            +--*  LCL_VAR   int    V02 cse0         u:1 $100
N002 (  1,  1) [000015] -----------                            \--*  CNS_INT   int    10 $42

------------ BB05 [0005] [016..018) (return), preds={BB04} succs={}

***** BB05 [0005]
STMT00005 ( 0x016[E-] ... 0x017 )
N002 (  2,  2) [000019] -----------                         *  RETURN    int    $VN.Void
N001 (  1,  1) [000024] -----------                         \--*  CNS_INT   int    0 $40

------------ BB06 [0004] [014..016) (return), preds={BB01,BB02,BB03,BB04} succs={}

***** BB06 [0004]
STMT00001 ( 0x014[E-] ... 0x015 )
N002 (  2,  2) [000005] -----------                         *  RETURN    int    $VN.Void
N001 (  1,  1) [000025] -----------                         \--*  CNS_INT   int    1 $46
```

JulieLeeMSFT commented:
> @JulieLeeMSFT regarding `return <conditions>` I have a prototype somewhere to normalize all `return <condition>` to if-else so all downstream phases don't have to care about it, so I presume we shouldn't care about BBJ_RETURN here for simplicity 🙂
Very good! Let's check that in.

> The `COMMA` in the first condition is something we can handle I guess here
Right, we didn't have time for that and left it for future work.
